File: javascript/food-chain/food-chain.js
const animals = [
  { name: 'fly', remark: null },
  { name: 'spider', remark: 'It wriggled and jiggled and tickled inside her.' },
  { name: 'bird', remark: 'How absurd to swallow a bird!' },
  { name: 'cat', remark: 'Imagine that, to swallow a cat!' },
  { name: 'dog', remark: 'What a hog, to swallow a dog!' },
  { name: 'goat', remark: 'Just opened her throat and swallowed a goat!' },
  { name: 'cow', remark: "I don't know how she swallowed a cow!" },
  { name: 'horse', remark: null }, // Horse has a special ending
];

const firstLine = (animal) => `I know an old lady who swallowed a ${animal}.\n`;
const lastLine = "I don't know why she swallowed the fly. Perhaps she'll die.\n";
const horseEnding = "She's dead, of course!\n";

const swallowLine = (currentAnimal, previousAnimal) => {
  let line = `She swallowed the ${currentAnimal} to catch the ${previousAnimal}`;
  if (previousAnimal === 'spider') {
    // The test suite expects the 'that wriggled...' part only when catching the spider from the bird verse onwards.
    // However, the example lyrics show it immediately in the spider verse. Let's adjust based on the example.
    // Re-reading the example: The spider verse itself has "It wriggled...", the *following* verses have "She swallowed the bird to catch the spider that wriggled..."
    // Let's stick to the logic derived from the full song example.
    line += ' that wriggled and jiggled and tickled inside her';
  }
  return line + '.\n';
};


export class Song {
  verse(n) {
    const index = n - 1;
    if (index < 0 || index >= animals.length) {
        throw new Error("Verse number out of range");
    }
    const animal = animals[index];
    let result = firstLine(animal.name);

    if (animal.name === 'horse') {
      result += horseEnding;
      return result;
    }

    if (animal.remark) {
      result += animal.remark + '\n';
    }

    // The first verse (fly) only has the first and last lines.
    if (animal.name === 'fly') {
        result += lastLine;
        return result;
    }

    // Build the cumulative part
    for (let i = index; i > 0; i--) {
      result += swallowLine(animals[i].name, animals[i - 1].name);
    }

    result += lastLine;
    return result;
  }

  verses(start, end) {
    if (start > end || start < 1 || end > animals.length) {
        throw new Error("Invalid verse range");
    }
    let result = '';
    for (let i = start; i <= end; i++) {
      result += this.verse(i);
      // Always add an extra newline for the blank line separation / final newline
      result += '\n';
    }
    return result;
  }
}

File: go/food-chain/food_chain.go
package foodchain

import "strings"

type animalInfo struct {
	name   string
	remark string
}

var animals = []animalInfo{
	{name: "fly", remark: "I don't know why she swallowed the fly. Perhaps she'll die."},
	{name: "spider", remark: "It wriggled and jiggled and tickled inside her."},
	{name: "bird", remark: "How absurd to swallow a bird!"},
	{name: "cat", remark: "Imagine that, to swallow a cat!"},
	{name: "dog", remark: "What a hog, to swallow a dog!"},
	{name: "goat", remark: "Just opened her throat and swallowed a goat!"},
	{name: "cow", remark: "I don't know how she swallowed a cow!"},
	{name: "horse", remark: "She's dead, of course!"},
}

// Verse generates a single verse of the song.
func Verse(v int) string {
	if v < 1 || v > len(animals) {
		return "" // Or handle error appropriately
	}
	idx := v - 1 // Adjust to 0-based index
	var sb strings.Builder

	// First line
	sb.WriteString("I know an old lady who swallowed a ")
	sb.WriteString(animals[idx].name)
	sb.WriteString(".\n")

	// Remark (except for fly, which is handled later, and horse)
	if idx > 0 && idx < len(animals)-1 {
		sb.WriteString(animals[idx].remark)
		sb.WriteString("\n")
	}

	// Special case for the last verse (horse)
	if idx == len(animals)-1 {
		sb.WriteString(animals[idx].remark)
		return sb.String()
	}

	// Cumulative lines
	for i := idx; i > 0; i-- {
		sb.WriteString("She swallowed the ")
		sb.WriteString(animals[i].name)
		sb.WriteString(" to catch the ")
		sb.WriteString(animals[i-1].name)
		// Special line for spider catching fly
		if i-1 == 1 { // animals[i-1] is spider
			sb.WriteString(" that wriggled and jiggled and tickled inside her")
		}
		sb.WriteString(".\n")
	}

	// Last line (remark for the fly)
	sb.WriteString(animals[0].remark)

	return sb.String()
}

// Verses generates a range of verses from start to end (inclusive).
func Verses(start, end int) string {
	if start < 1 || end > len(animals) || start > end {
		return "" // Or handle error
	}
	var sb strings.Builder
	for i := start; i <= end; i++ {
		sb.WriteString(Verse(i))
		if i < end {
			sb.WriteString("\n\n") // Add blank line between verses
		}
	}
	return sb.String()
}

// Song generates the entire song.
func Song() string {
	return Verses(1, len(animals))
}
